### PR TITLE
feat: Add subscription form to email report

### PR DESCRIPTION
**Pull request by wa0x6e**
### Issues

Fixes https://github.com/snapshot-labs/envelop/issues/35

This PR add a form to subscribe to snapshot email notifications, from the user account menu on the header

### Changes 

1. Add email form to subscribe to envelop email notifications 

### How to test

1. Log in to your account
2. Click on your profile in the header navigation
3. In the dropdown, select "SUBSCRIBE"
4. Fill the email form in the modal
5. Submit
6. Sign the message
7. See success/error message

### To-Do

- [x] Show the form in relevant places 
- [x] Add modal/message when subscription is successful
- [x] Add error message on error

### Self-review checklist
- [x] I have performed a full self-review of my changes
- [x] I have tested my changes on a preview deployment
- [ ] I have tested my changes on a custom domain
- [ ] I have run end-to-end tests `yarn cypress:test:e2e`, and they have passed

### CODE REVIEW

**Comment by samuveth** [comment]
Would be create if you could add a Loom recording too

**Comment by wa0x6e** [comment]
Here you go https://www.loom.com/share/85bd767defd74f96b08dc2b11705b850

Workflow lead to an error message currently, as the backend server is offline

**Comment by wa0x6e** [comment]
@samuveth , please take a look to the code requiring the user signature, proving he's the wallet owner.

It works, but I am not sure it fits the coding philosophy. I looked around, and I think that everything signature and verification related logic are delegated to the snapshot SDK.

The new composable introduced is using a sign function directly from web3, and it seems to be the only piece of code to to that. SDK doesn't seem to expose a function to only sign, without contacting the sequencer.

File: src/composables/useEmailSubscription.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
+import { getInstance } from '@snapshot-labs/lock/plugins/vue3';
+
+export function useEmailSubscription() {
+  enum Status {
+    wating,
```
Comment (samuveth) [inline src/composables/useEmailSubscription.ts:5]: Typo here, should be `waiting` I think

File: src/composables/useEmailSubscription.ts [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+  async function sign(
+    message: {
+      email: string;
+      from: string;
+      timestamp: number;
+    },
+    types: any
+  ) {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
```
Comment (samuveth) [inline src/composables/useEmailSubscription.ts:37]: Why is this needed, I don't see any errors if I remove it

File: src/locales/default.json [hunk 3]
```diff
@@ -892,6 +892,16 @@
     "title": "Get the latest Snapshot updates",
     "join": "Join Snapshot newsletter"
   },
+  "emailSubscription": {
+    "title": "Email subscription",
+    "subscribe": "Subscribe",
+    "description": "Subscribe to receive email notifications about your joined spaces and proposals activities.",
+    "postSubscribeMessage": {
+      "success": "You have been subscribed to email notifications. You will receive an email shortly with a verification link in order to confirm your subscription.",
+      "apiError": "An error occured while adding your email to the subscriptions list",
```
Comment (samuveth) [inline src/locales/default.json:901]:
```suggestion
      "apiError": "An error ocurred while adding your email to the subscriptions list",
```

**Comment by samuveth** [comment]
> @samuveth , please take a look to the code requiring the user signature, proving he's the wallet owner.
> 
> It works, but I am not sure it fits the coding philosophy. I looked around, and I think that everything signature and verification related logic are delegated to the snapshot SDK.
> 
> The new composable introduced is using a sign function directly from web3, and it seems to be the only piece of code to to that. SDK doesn't seem to expose a function to only sign, without contacting the sequencer.

Can't we use the `sign` method on the client in `helpers/clientEIP712`?

**Comment by samuveth** [comment]
Why do we need two buttons?
image

**Comment by samuveth** [comment]
We should consider incorporating a subscription modal following the voting process. After a user submits their vote and closes the post-vote modal, if they are not already subscribed, a new modal could appear prompting them to subscribe. This approach will help engage users and encourage subscriptions in a seamless manner, particularly since the current dropdown option for subscribing isn't very prominent and may be easily overlooked.

**Comment by wa0x6e** [comment]
> 

2 modals in a row, isn't that too much ? I think there are already some 'shares' button on the post vote modal, maybe just add another for 'notifiy me by email' ?

File: src/composables/useEmailSubscription.ts (hunk 2 above)
Comment (wa0x6e) [reply src/composables/useEmailSubscription.ts:37]: the use of @ts-ignore raises an error

**Comment by wa0x6e** [comment]
sing extract to helper, and made some UI improvement

File: src/components/ModalEmailSubscription.vue [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    </BaseMessageBlock>
+    <div v-else class="m-4">
+      {{ $t('emailSubscription.description') }}
+    </div>
+    <form v-if="status !== Status.success" class="m-4" @submit.prevent="submit">
+      <BaseInput
+        v-bind="$attrs"
+        :model-value="email"
+        type="email"
+        :placeholder="'johndoe@gmail.com'"
```
Comment (Todmy) [inline src/components/ModalEmailSubscription.vue:43]: can be replaced to 
```suggestion
        placeholder="johndoe@gmail.com"
```

File: src/components/ModalEmailSubscription.vue [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        autocomplete="off"
+        @update:model-value="emit('update:modelValue', $event)"
+      >
+        <template #before>
+          <i-ho-mail class="text-[16px]" />
+        </template>
+      </BaseInput>
+
+      <small>
+        You may be asked to sign a transaction to verify wallet ownership
```
Comment (Todmy) [inline src/components/ModalEmailSubscription.vue:56]: do we need a translation for this?

**Review by Todmy — CHANGES_REQUESTED** [review]

File: src/helpers/sign.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+export default async function sign(
+  web3: Web3Provider | Wallet,
+  address: string,
+  message: ISubscribe,
+  types: DataType
+) {
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  const signer = web3?.getSigner ? web3.getSigner() : web3;
```
Comment (samuveth) [inline src/helpers/sign.ts:33]:
```js
  let signer: any;
  if ('getSigner' in web3) {
    signer = web3.getSigner();
  } else {
    signer = web3;
  }
  ```
  
  Suggest this instead of using ts-ignore

File: src/components/ModalEmailSubscription.vue (hunk 5 above)
Comment (samuveth) [reply src/components/ModalEmailSubscription.vue:56]: No, this should say `sign message` instead

**Comment by samuveth** [comment]
> 2 modals in a row, isn't that too much ? I think there are already some 'shares' button on the post vote modal, maybe just add another for 'notifiy me by email' ?

Yes, that could also work. Can you add that?

**Comment by wa0x6e** [comment]
@Todmy will take over PR from now, as he's the one in charge of the front-end side of envelop. I will go into reviewer mode.

**Comment by aurelianoB** [comment]
> > 2 modals in a row, isn't that too much ? I think there are already some 'shares' button on the post vote modal, maybe just add another for 'notifiy me by email' ?
> 
> Yes, that could also work. Can you add that?

In favour of this as well. Is this already contemplated in some issue, or should we create a specific one for this?

**Comment by samuveth** [comment]
> > > 2 modals in a row, isn't that too much ? I think there are already some 'shares' button on the post vote modal, maybe just add another for 'notifiy me by email' ?
> > 
> > 
> > Yes, that could also work. Can you add that?
> 
> In favour of this as well. Is this already contemplated in some issue, or should we create a specific one for it?

I think we will add it inside this PR directly

File: src/components/ModalEmailSubscription.vue (hunk 5 above)
Comment (samuveth) [reply src/components/ModalEmailSubscription.vue:56]: Oh sorry, yes we should add it from defauls.json but down worry about the translation files these are handles through Crowdin.

File: src/locales/default.json [hunk 7]
```diff
@@ -910,6 +910,7 @@
     "title": "Email subscription",
     "subscribe": "Subscribe",
     "description": "Subscribe to receive email notifications about your joined spaces and proposals activities.",
+    "buttonCaption": "Sign message",
```
Comment (samuveth) [inline src/locales/default.json:913]: What I mean was it should be `sign a message` instead of `sign a transaction`

**Comment by samuveth** [comment]
You're right, I think the current success message is too long. We can shorten it and add a checkmark for immediate success recognition.

**Comment by samuveth** [comment]
Some suggestions for a shorter text:

```
"Subscription successful! Please confirm via the verification link sent to your email."

"Thanks for subscribing! A confirmation link has been sent to your email. Click to verify."

"You're subscribed! Please confirm by clicking the verification link in your email."
```

**Comment by aurelianoB** [comment]
> ```
> "Thanks for subscribing! A confirmation link has been sent to your email. Click to verify."
> ```

I'd go for this one, as the others suggest subscription is complete, which is not the case.

**Comment by Todmy** [comment]
@samuveth @wa0x6e @aurelianoB let me clarify. We need to modify the `success subscription` modal (for both cases `subscription after vote` and `subscription from the menu`) by shortening the text to `Thanks for subscribing! A confirmation link has been sent to your email. Click to verify.` and add a big checkmark image above.
Do I get you right?

**Comment by Todmy** [comment]
Screenshot 2023-05-10 at 10 52 06
What do you think?

**Comment by aurelianoB** [comment]
looks good @Todmy - minor detail, but maybe you can use a green checkmark

**Comment by wa0x6e** [comment]
I was visualizing something in this style, for the text formatting.

Screenshot 2023-04-04 at 11 47 03 AM

We remove the share buttons, and maybe replace the picture by a bigger checkmark icon, but not sure that the checkmark icon will be in line with the current design guideline

**Comment by Todmy** [comment]
This is how subscription success modal looks like:

Screenshot 2023-05-10 at 17 27 41

File: src/locales/default.json [hunk 8]
```diff
@@ -256,7 +256,8 @@
       "seeQueue": "See queued transactions",
       "tips": {
         "1": "Votes can be changed while the proposal is active"
-      }
+      },
+      "subscribe": "Subscribe with Email"
```
Comment (samuveth) [inline src/locales/default.json:260]:
```suggestion
      "subscribe": "Subscribe with email"
```

**Review by samuveth — APPROVED** [review]
tACK

**Comment by wa0x6e** [comment]
2 concepts of the success modal UI

Screenshot 2023-05-15 at 11 14 03 PM
Screenshot 2023-05-15 at 11 13 28 PM

If approved, I will commit them

**Comment by aurelianoB** [comment]
I'd vote for the first one!

**Comment by aurelianoB** [comment]
This is the current error message I see when rejecting message signature - what different message would you show?

image

**Comment by aurelianoB** [comment]
Something more concise could be "Sign message to verify wallet ownership"

**Comment by wa0x6e** [comment]
Or maybe: "Please sign the message to proceed. The signature is required to verify wallet ownership and prevent spam".

We should tell what went wrong, how to fix, and why it's required, in the more concise way.
